lib/spack/spack/build_systems/intel.py

```diff
@@ -0,0 +1,190 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+
+import os
+import xml.etree.ElementTree as ET
+
+from llnl.util.filesystem import install, join_path
+from spack.package import PackageBase, run_after
+from spack.util.executable import Executable
+
+
+def _valid_components():
+    """A generator that yields valid components."""
+
+    tree = ET.parse('pset/mediaconfig.xml')
+    root = tree.getroot()
+
+    components = root.findall('.//Abbr')
+    for component in components:
+        # For some reason, mediaconfig.xml contains an invalid component:
+        if component.text != 'intel-tbb-psxe__noarch':
+            yield component.text
+
+
+class IntelPackage(PackageBase):
+    """Specialized class for licensed Intel software.
+
+    This class provides two phases that can be overridden:
+
+    1. :py:meth:`~.IntelPackage.configure`
+    2. :py:meth:`~.IntelPackage.install`
+
+    They both have sensible defaults and for many packages the
+    only thing necessary will be to override ``setup_environment``
+    to set the appropriate environment variables.
+    """
+    #: Phases of an Intel package
+    phases = ['configure', 'install']
+
+    #: This attribute is used in UI queries that need to know the build
+    #: system base class
+    build_system_class = 'IntelPackage'
+
+    #: By default, we assume that all Intel software requires a license.
+    #: This can be overridden for packages that do not require a license.
+    license_required = True
+
+    #: Comment symbol used in the ``license.lic`` file
+    license_comment = '#'
+
+    #: Location where Intel searches for a license file
+    license_files = ['Licenses/license.lic']
+
+    #: Environment variables that Intel searches for a license file
+    license_vars = ['INTEL_LICENSE_FILE']
+
+    #: URL providing information on how to acquire a license key
+    license_url = 'https://software.intel.com/en-us/articles/intel-license-manager-faq'
+
+    #: Components of the package to install.
+    #: By default, install 'ALL' components.
+    components = ['ALL']
+
+    @property
+    def _filtered_components(self):
+        """Returns a list of valid components that match the requested
+        components from ``components``."""
+
+        # Don't filter 'ALL'
+        if self.components == ['ALL']:
+            return self.components
+
+        matches = []
+
+        for valid in _valid_components():
+            for requested in self.components:
+                if valid.startswith(requested):
+                    matches.append(valid)
+
+        return matches
```

Comment:
if we "from sets import Set" and "return Set(matches)", then I think this will avoid duplication of components.